proxy/envoy/egress.go

```diff
@@ -0,0 +1,197 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package envoy
+
+import (
+	"fmt"
+	"sort"
+	"time"
+
+	"github.com/golang/glog"
+
+	"istio.io/api/proxy/v1/config"
+	"istio.io/manager/model"
+	"istio.io/manager/proxy"
+)
+
+type egressWatcher struct {
+	agent   proxy.Agent
+	ctl     model.Controller
+	context *EgressConfig
+}
+
+// NewEgressWatcher creates a new egress watcher instance with an agent
+func NewEgressWatcher(ctl model.Controller, context *EgressConfig) (Watcher, error) {
+	agent := proxy.NewAgent(runEnvoy(context.Mesh, "egress"), cleanupEnvoy(context.Mesh), 10, 100*time.Millisecond)
+
+	out := &egressWatcher{
+		agent:   agent,
+		ctl:     ctl,
+		context: context,
+	}
+
+	// egress depends on the external services declaration being up to date
+	if err := ctl.AppendServiceHandler(func(*model.Service, model.Event) {
+		out.reload()
+	}); err != nil {
+		return nil, err
+	}
+
+	return out, nil
+}
+
+func (w *egressWatcher) reload() {
+	w.agent.ScheduleConfigUpdate(generateEgress(w.context))
+}
+
+func (w *egressWatcher) Run(stop <-chan struct{}) {
+	go w.agent.Run(stop)
+
+	// Initialize envoy according to the current model state,
+	// instead of waiting for the first event to arrive.
+	// Note that this is currently done synchronously (blocking),
+	// to avoid racing with controller events lurking around the corner.
+	// This can be improved once we switch to a mechanism where reloads
+	// are linearized (e.g., by a single goroutine reloader).
+	w.reload()
+	w.ctl.Run(stop)
+}
+
+// EgressConfig defines information for engress
+type EgressConfig struct {
+	Namespace string
+	Services  model.ServiceDiscovery
+	Mesh      *config.ProxyMeshConfig
+}
+
+func generateEgress(conf *EgressConfig) *Config {
+
+	// Create a VirtualHost for each external service
+	vhosts := make([]*VirtualHost, 0)
+	services := conf.Services.Services()
+	for _, service := range services {
+		if service.External {
+			vhosts = append(vhosts, buildEgressHTTPRoute(service))
+		}
+	}
+
+	sort.Slice(vhosts, func(i, j int) bool { return vhosts[i].Name < vhosts[j].Name })
+
+	rConfig := &HTTPRouteConfig{VirtualHosts: vhosts}
+
+	listener := &Listener{
+		Address:    fmt.Sprintf("tcp://%s:80", WildcardAddress),
+		BindToPort: true,
+		Filters: []*NetworkFilter{
+			{
+				Type: "read",
+				Name: HTTPConnectionManager,
+				Config: HTTPFilterConfig{
+					CodecType:   "auto",
+					StatPrefix:  "http",
+					AccessLog:   []AccessLog{{Path: DefaultAccessLog}},
+					RouteConfig: rConfig,
+					Filters: []HTTPFilter{
+						{
+							Type:   "decoder",
+							Name:   "router",
+							Config: FilterRouterConfig{},
+						},
+					},
+				},
+			},
+		},
+	}
+
+	//TODO
```

Comment:
Doesn't apply to egress right? Please remove comment. But add a comment related to tcp service